crates/stages/stages/src/stages/merkle_changesets.rs

```diff
@@ -0,0 +1,374 @@
+        );
+
+        // We need to distinguish a full revert and a per-block revert. A full revert reverts
+        // changes starting at db tip all the way to a block. A per-block revert only reverts
+        // a block's changes.
+        //
+        // We need to calculate the full HashedPostState reverts for every block in the target
+        // range. The full HashedPostState revert for block N can be calculated as:
+        //
+        //
+        // ```
+        // // where `extend` overwrites any shared keys
+        // state_revert(N) = state_revert(N + 1).extend(per_block_state_revert(N))
+        // ```
+        //
+        // We need per-block reverts to calculate the prefix set for each individual block. By using
+        // the per-block reverts to calculate full reverts on-the-fly we can save a bunch of memory.
+        debug!(
+            target: "sync::stages::merkle_changesets",
+            ?target_range,
+            "Computing per-block state reverts",
+        );
+        let mut per_block_state_reverts = Vec::new();
+        for block_number in target_range.clone() {
+            per_block_state_reverts.push(HashedPostState::from_reverts::<KeccakKeyHasher>(
+                provider.tx_ref(),
+                block_number..=block_number,
+            )?);
+        }
+
+        let per_block_state_revert = |block_number| -> &HashedPostState {
+            &per_block_state_reverts[(block_number - target_start) as usize]
+        };
+
+        let state_revert = |block_number| -> HashedPostState {
+            let mut r = HashedPostState::default();
+            for n in (block_number..target_end).rev() {
+                r.extend_ref(per_block_state_revert(n))
+            }
+            r
+        };
+
+        // To calculate the changeset for a block, we first need the TrieUpdates which are generated
+        // as a result of processing the block. To get these we need:
+        // 1) The TrieUpdates which revert the db's trie to _prior_ to the block
+        // 2) The HashedPostState to revert the db's state to _after_ the block
+        //
+        // To get (1) for `target_start` we need to do a big state root calculation which takes into
+        // account all changes between that block and db tip. For each block after the
+        // `target_start` we can update (1) using the TrieUpdates which were output by the previous
+        // block only targeting the state changes of that block.
+        debug!(
+            target: "sync::stages::merkle_changesets",
+            ?target_start,
+            "Computing trie state at starting block",
+        );
+        let mut input = TrieInput::default();
+        input.state = state_revert(target_start);
+        input.prefix_sets = input.state.construct_prefix_sets();
+        // target_start will be >= 1, see `determine_target_range`.
+        input.nodes =
+            Self::calculate_block_trie_updates(provider, target_start - 1, input.clone())?;
```

Comment:
can introduce a method `TrieInput::from_state_and_nodes` or smth that will be similar to `TrieInput::from_state`